README
NAME
    Test::Version - Check to see that version's in modules are sane

VERSION
    version 1.002001

SYNOPSIS
            use Test::More;
            use Test::Version 1.001001 qw( version_all_ok ), {
                            is_strict   => 0,
                            has_version => 1,
                    };

            # test blib or lib by default
            version_all_ok();

            done_testing;

DESCRIPTION
    This module's goal is to be a one stop shop for checking to see that
    your versions across your dist are sane. Please ensure that you use
    version 0.04 or later only, as earlier versions are old code and may not
    work correctly. Current feature list:

    module has a version
        Tests to insure that all modules checked have a VERSION defined, Can
        replace Test::HasVersion

    module has a valid version
        Tests to insure that all versions are valid, according to the rules
        of version method "is_lax". To quote:

        *The lax criteria corresponds to what is currently allowed by the
        version parser. All of the following formats are acceptable for
        dotted-decimal formats strings:*

                v1.2
                1.2345.6
                v1.23_4
                1.2345
                1.2345_01

        *If you want to limit yourself to a much more narrow definition of
        what a version string constitutes, is_strict() is limited to version
        strings like the following list:*

                v1.234.5
                2.3456

        you can cause your tests to fail if not strict by setting is_strict
        to 1

FUNCTIONS
  version_ok
            version_ok( $filename, [ $name ] );

    Test a single ".pm" file by passing a path to the function. Checks if
    the module has a version, and that it is valid with "is_lax".

  version_all_ok
            version_all_ok( [ $directory, [ $name ]] );

    Test all modules in a directory with "version_ok". By default it will
    check "blib" or "lib" if you haven't passed it a directory.

CONFIGURATION AND ENVIRONMENT
  has_version
            use Test::Version qw( version_all_ok ), { has_version => 0 };

    Allows disabling whether a module has to have a version. If set to 0
    version tests will be skipped in any module where no version is found.

    really doesn't make sense to use with just version_ok

  is_strict
            use Test::Version { is_strict => 1 };

    this allows enabling of versions "is_strict" checks to ensure that your
    version is strict.

LIMITATIONS
    Will not test Perl 5.12 "package" version declarations because
    Module::Extract::VERSION can't extract them yet.

SEE ALSO
    The goal is to have the functionality of all of these.

    Test::HasVersion
    Test::ConsistentVersion
    Test::GreaterVersion

ACKNOWLEDGEMENTS
    Special thanks to particle "<particle at cpan dot org"> for the original
    "Test::Version" and letting me maintain it further. Thanks to Mike
    Doherty "<doherty at cs dot dal dot ca">, and Michael G. Schwern
    "<schwern at pobox dot com"> for their patches.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/xenoterracide/Test-Version/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Caleb Cushing <xenoterracide@gmail.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2011 by Caleb Cushing.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

1.004001  2014-10-20
	- improved consistent check diagnostics GH #11 (SZABGAB)

1.004000  2014-10-18
	- add consistent check GH #10 (SZABGAB)

1.003001  2014-09-06
	- remove inline and remove _get_version, trying to skip test there won't
	  work GH #4 (ETHER)

1.003000  2014-09-05
	- skip packages unindexable by pause GH #4 (ETHER)

1.002004  2013-11-21
	- fix bugs in argument handling
	- fix whitespace

1.002003  2013-10-14
	- fix misgithap
	- more dist.ini updates

1.002002  2013-10-14
	- fix synopsis GH #6, #7 ( plicease )
	- change Dist::Zilla Plugins
	- remove old documentation that no longer applies

1.002001  2012-03-14
	- fix metadata caused by a bug in DZP::GitHub after asking repo to be
	  unlinked from gitpan.

1.002000  2012-01-09
	- use Module::Metadata
	- allow disabling of 'has_version'
	- require at least 1 version
	- allow for checking that a module is_strict

1.001003  2011-12-01
	- use Module::Metadata apparently it's closer to how Perl works than
	  Module::Extract::VERSION

1.001002  2011-11-02
	- fix some issues in the pod

1.001001  2011-11-01
	- provide is_strict setting to replace STRICTNESS
	- provide has_version setting to disable that a module must always have a
	  version
	- use perl decimal style semantic versioning because of bugs in EUMM with
	  vstring versions

v1.1.0    2011-08-14
	- check if versions are strict
	- provide STRICTNESS settings to decide what to do if they aren't strict

1.0.0     2011-07-01
	- Change to Semantic Versioning
	- Consider API Stable
	- No real changes since version 0.11

0.11      2011-06-18
	- correct the version of version.pm that we need to depend on. is_lax was
	  not properly exported until version 0.86

0.10      2011-06-08
	- fix warning on bad api call #1
	- make diagnostics human readable #1
	- remove diagnostics from pod #1
	- change return values #1
	- improve tests #1

0.09      2011-06-07
	- Test for FILE_NOT_DEFINED was not actually included in the last release.
	  Add it.

0.08      2011-06-04
	- test NO_FILE and FILE_NOT_DEFINED
	- document only use 0.04 or later
	- fix 04-all.t test that compares the arrays

0.07      2011-05-29
	- more documentation
	- don't use diag for VERSION_OK status

0.06      2011-05-28
	- update diag messages
	- check that $file exists
	- switch to Test::Tester exclusively
	- fix 04-all.t from failing due to array order being random

0.05      2011-05-27
	- switch to Test::Tester for some tests.
	- update changelog for 0.04 due to leaving featurelist out

0.04      2011-05-26
	- Release new codebase as a trial version
	- Ability to iterate through a directory checking versions with
	  version_all_ok()
	- validate versions with is_lax

0.03      2011-05-26
	- Maintainership passed to Caleb Cushing <xenoterracide@gmail.com>
	- Deprecate "expected" parameter to version_ok
	- Switch to Dist::Zilla for release management

0.02	2002-11-18
	- removed rogue 'plan' statement in Version.pm,
	  which was causing test failure

0.01	2002-11-17
	- original version; created by h2xs 1.21 with options
		-AX Test::Version



lib/Test/Version.pm
package Test::Version;
use 5.006;
use strict;
use warnings;
use Carp;

# VERSION

use parent 'Exporter';
use Test::Builder;
use version 0.86 qw( is_lax is_strict );
use File::Find::Rule::Perl;
use Test::More;
use Module::Metadata;

our @EXPORT = qw( version_all_ok ); ## no critic (Modules::ProhibitAutomaticExportation)
our @EXPORT_OK = qw( version_ok );

my $cfg;

sub import { ## no critic qw( Subroutines::RequireArgUnpacking Subroutines::RequireFinalReturn )
	my @exports;
	foreach my $param ( @_ ) {
		unless ( ref( $param ) eq 'HASH' ) {
			push @exports, $param;
		} else {
			$cfg = $param
		}
	}

	$cfg->{is_strict}
		= defined $cfg->{is_strict}           ? $cfg->{is_strict}
		:                                       0
		;

	$cfg->{has_version}
		= defined $cfg->{has_version}         ? $cfg->{has_version}
		:                                       1
		;

	$cfg->{consistent}
		= defined $cfg->{consistent}           ? $cfg->{consistent}
		:                                       0
		;

	my $mmv = version->parse( $Module::Metadata::VERSION );
	my $rec = version->parse( '1.000020'  );
	if ( $mmv >= $rec && ! defined $cfg->{ignore_unindexable} ) {
		$cfg->{ignore_unindexable} = 1;
	}

	__PACKAGE__->export_to_level( 1, @exports );
}

my $version_counter = 0;
my $version_number;
my $consistent = 1;
my %versions;

my $test = Test::Builder->new;

sub version_ok {
	my ( $file, $name ) = @_;
	$file ||= '';
	$name ||= "check version in '$file'";

	croak 'No file passed to version_ok().' unless $file;

	croak "'$file' doesn't exist." unless -e $file;

	my $info = Module::Metadata->new_from_file( $file );
	if ( $cfg->{ignore_unindexable} ) {
		$test->skip( "$file not indexable" );
		return 0 if ! $info->is_indexable;
	}
	my $version = $info->version;

	$versions{$file} = $version;

	if (not defined $version) {
		$consistent = 0;
	}

	if ( not $version and not $cfg->{has_version} ) {
		$test->skip( 'No version was found in "'
			. $file
			. '" and has_version is false'
			)
			;

		return 1;
	} else {
		$version_counter++;
	}

	unless ( $version ) {
		$test->ok( 0 , $name );
		$test->diag( "No version was found in '$file'." );
		return 0;
	}

	unless (defined $version_number) {
		$version_number = $version;
	}
	if ($version ne $version_number) {
		$consistent = 0;
	}

	unless ( is_lax( $version ) ) {
		$test->ok( 0, $name );
		$test->diag( "The version '$version' found in '$file' is invalid." );
		return 0;
	}

	if ( $cfg->{is_strict} ) {
		unless ( is_strict( $version ) ) {
			$test->ok( 0, $name );
			$test->diag( "The version '$version' found in '$file' is not strict." );
			return 0;
		}
	}

	$test->ok( 1, $name );
	return 1;
}

sub version_all_ok {
	my ( $dir, $name ) = @_;

	$dir
		= defined $dir ? $dir
		: -d 'blib'    ? 'blib'
		:                'lib'
		;

	croak $dir . ' does not exist, or is not a directory' unless -d $dir;

	# Report failure location correctly - GH #1
	local $Test::Builder::Level = $Test::Builder::Level + 1; ## no critic (Variables::ProhibitPackageVars)

	$name ||= "all modules in $dir have valid versions";

	my @files = File::Find::Rule->perl_module->in( $dir );

	foreach my $file ( @files ) {
		version_ok( $file );
	}

	if ($cfg->{consistent} and not $consistent) {
		$test->ok( 0, $name );
		$test->diag('The version numbers in this distribution are not the same');
		foreach my $file (sort { $versions{$a} cmp $versions{$b} } keys %versions) {
			$test->diag(sprintf "%10s %s", (defined $versions{$file} ? $versions{$file} : 'undef'), $file);
		}
		return;
	}

	# has at least 1 version in the dist
	if ( not $cfg->{has_version} and $version_counter < 1 ) {
		$test->ok( 0, $name );
		$test->diag(
			'Your dist has no valid versions defined. '
			. 'Must have at least one version'
			);
	}
	else {
		$test->ok( 1, $name );
	}

	return;
}
1;

# ABSTRACT: Check to see that version's in modules are sane

=head1 SYNOPSIS

	use Test::More;
	use Test::Version 1.001001 qw( version_all_ok ), {
			is_strict   => 0,
			has_version => 1,
			consistent  => 1,
		};

	# test blib or lib by default
	version_all_ok();

	done_testing;

=head1 DESCRIPTION

This module's goal is to be a one stop shop for checking to see that your
versions across your dist are sane. Please ensure that you use version C<0.04>
or later only, as earlier versions are old code and may not work correctly.
Current feature list:

=over

=item module has a version

Tests to insure that all modules checked have a VERSION defined, Can replace
L<Test::HasVersion>

=item module has a valid version

Tests to insure that all versions are valid, according to the rules of
L<version> method C<is_lax>. To quote:

I<The lax criteria corresponds to what is currently allowed by the version
parser. All of the following formats are acceptable for dotted-decimal formats
strings:>

	v1.2
	1.2345.6
	v1.23_4
	1.2345
	1.2345_01

I<If you want to limit yourself to a much more narrow definition of what a
version string constitutes, is_strict() is limited to version strings like
the following list:>

	v1.234.5
	2.3456

you can cause your tests to fail if not strict by setting L<is_strict|/is_strict> to
C<1>

=back

=function version_ok

	version_ok( $filename, [ $name ] );

Test a single C<.pm> file by passing a path to the function. Checks if the
module has a version, and that it is valid with C<is_lax>.

=function version_all_ok

	version_all_ok( [ $directory, [ $name ]] );

Test all modules in a directory with C<version_ok>. By default it will check
C<blib> or C<lib> if you haven't passed it a directory.

=setting has_version

	use Test::Version qw( version_all_ok ), { has_version => 0 };

Allows disabling whether a module has to have a version. If set to 0
version tests will be skipped in any module where no version is found.

really doesn't make sense to use with just L<version_ok|/version_ok>

=setting is_strict

	use Test::Version { is_strict => 1 };

this allows enabling of L<version>s C<is_strict> checks to ensure that your
version is strict.

=setting consistent

	use Test::Version { consistent => 1 };

Check if every module has the same version number.

=setting ignore_unindexable

	use Test::Version { ignore_unindexable => 0};

if you have at least L<Module::Metadata> vC<1.000020> Test::Version will by
default skip any files not considered L<is_indexable|Module::Metadata/is_indexable>

=head1 SEE ALSO

The goal is to have the functionality of all of these.

=over

=item L<Test::HasVersion>

=item L<Test::ConsistentVersion>

=item L<Test::GreaterVersion>

=back

=cut


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Check to see that version's in modules are sane",
  "AUTHOR" => "Caleb Cushing <xenoterracide\@gmail.com>",
  "BUILD_REQUIRES" => {
    "English" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Scalar::Util" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.94",
    "Test::Requires" => 0,
    "Test::Tester" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Test-Version",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Test::Version",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "File::Find::Rule::Perl" => 0,
    "Module::Metadata" => 0,
    "Test::Builder" => 0,
    "Test::More" => 0,
    "parent" => 0,
    "strict" => 0,
    "version" => "0.86",
    "warnings" => 0
  },
  "VERSION" => "1.002001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



